rasa/shared/core/events.py

```diff
@@ -1210,7 +1210,9 @@ def as_sub_state(self) -> Dict[Text, Text]:
         if self.action_name:
             return {ACTION_NAME: self.action_name}
         else:
-            return {ACTION_TEXT: self.action_text}
+            # FIXME: we should define the type better here, and require either
+            #        `action_name` or `action_text`
+            return {ACTION_TEXT: cast(Text, self.action_text)}
```

Comment:
Why is this a problem? We return a `Dict[Text, Optional[Text]]` either way

Reply:
I think it's a problem because an `ActionExecuted` instance should either have `action_text` or `action_name` defined (not both at the same time nor none), as I understand it (let me know if this is wrong).

I added the cast because the type definition of this function and others (including `DialogueStateTracker.set_latest_action()`) expect the type to be `Dict[Text, Text]`, not `Dict[Text, Optional[Text]]`.

Reply:
Yes, it's always either.  What do you think of doing something like:
```
        if self.action_name:
            return {ACTION_NAME: self.action_name}
        elif self.action_text:
            return {ACTION_TEXT: self.action_text}
       else:
           raise ValueError("{self.class.__name} does neither have {ACTION_NAME} nor {ACTION_TEXT} specified. Please check if your training data is valid."
```

This exception should never come up. However, this makes it fail loud and clear.

Reply:
yes I wanted to do that as well, but it won't fix the typing issue I'm afraid, as it's unlikely that mypy will pick it up. I'll try though